cosmos/x/evm/plugins/precompile/plugin.go

```diff
@@ -119,41 +120,69 @@ func (p *plugin) SetTransientKVGasConfig(transientKVGasConfig storetypes.GasConf
 // Run implements core.PrecompilePlugin.
 func (p *plugin) Run(
 	evm ethprecompile.EVM, pc vm.PrecompileContainer, input []byte,
-	caller common.Address, value *big.Int, suppliedGas uint64, readonly bool,
+	caller common.Address, value *big.Int, suppliedGas uint64, readOnly bool,
 ) ([]byte, uint64, error) {
 	// use a precompile-specific gas meter for dynamic consumption
 	gm := storetypes.NewInfiniteGasMeter()
 	// consume static gas from RequiredGas
 	gm.ConsumeGas(pc.RequiredGas(input), "RequiredGas")
 
-	// get native Cosmos SDK context from the Polaris StateDB
+	// get native Cosmos SDK context and MultiStore from the Polaris StateDB
 	sdb := utils.MustGetAs[vm.PolarisStateDB](evm.GetStateDB())
 	ctx := sdk.UnwrapSDKContext(sdb.GetContext())
+	ms := utils.MustGetAs[MultiStore](ctx.MultiStore())
+
+	// make sure the readOnly is only set if we aren't in readOnly yet, which also makes sure that
+	// the readOnly flag isn't removed for child calls (taken from geth core/vm/interepreter.go)
+	if readOnly && !ms.IsReadOnly() {
+		ms.SetReadOnly(true)
+		defer func() { ms.SetReadOnly(false) }()
+	}
 
 	// disable reentrancy into the EVM
 	p.disableReentrancy(sdb)
 
-	// run precompile container
-	ret, err := pc.Run(
+	// recover from any panic during precompile execution for the EVM to handle as a vm error
+	var err error
+	defer func() {
+		if panicked := recover(); panicked != nil {
+			// NOTE: this only propagates an error back to the EVM if the type of the given panic
+			// value is error or string (any other type of panic value is ignored)
+			var isError bool
+			if err, isError = utils.GetAs[error](panicked); !isError {
+				if strErr, isString := utils.GetAs[string](panicked); isString {
+					err = errors.New(strErr)
+				}
+			}
+		}
+	}()
+
+	// run the precompile container
+	var ret []byte
+	ret, err = pc.Run(
 		ctx.WithGasMeter(gm).
 			WithKVGasConfig(p.kvGasConfig).
 			WithTransientKVGasConfig(p.transientKVGasConfig),
 		evm,
 		input,
 		caller,
 		value,
-		readonly,
+		readOnly,
 	)
 
 	// enable reentrancy into the EVM
 	p.enableReentrancy(sdb)
 
 	// handle overconsumption of gas
 	if gm.GasConsumed() > suppliedGas {
-		return nil, 0, vm.ErrOutOfGas
+		suppliedGas = gm.GasConsumed() // ensures 0 gas remaining if gas was overconsumed
+		if err == nil {
+			// only return out of gas error if no other error occurred
+			err = vm.ErrOutOfGas
+		}
```

Comment:
should out of gas error take priority over all other errors? imo its more accurate to return the other errors (like ExecutionReverted, WriteProtection, etc.) bc the evm specifically handles these

Reply:
what does geth do?

Reply:
geth consumes gas per opcode, so if it hits out of gas after any opcode, it will just return it. we wait till the end of the execution to check for overconsumption of gas